pkg/lib/cockpit-components-terminal.tsx

```diff
@@ -120,71 +226,93 @@ export class Terminal extends React.Component {
-                }
-                this.props.channel.send(data);
-            }
-        }.bind(this));
+        this.terminalRef = React.createRef<HTMLDivElement>();
 
         if (props.onTitleChanged)
             term.onTitleChange(props.onTitleChanged);
 
         this.terminal = term;
         this.state = {
             showPastingModal: false,
-            cols: props.cols || 80,
-            rows: props.rows || 25
+            cols: term.cols,
+            rows: term.rows
         };
     }
 
-    componentDidMount() {
-        this.terminal.open(this.terminalRef.current);
-        this.terminal.loadAddon(new CanvasAddon());
-        this.connectChannel();
+    mountTerminal(state: TerminalState) {
+        this.terminal = state.terminal;
+        this.terminalRef.current?.appendChild(state.element);
+        this.terminal.open(state.element);
+
+        if (this.props.fontSize)
+            this.terminal.options.fontSize = this.props.fontSize;
+
+        if (this.props.cols && this.props.rows) {
+            this.resizeTerminal(this.props.cols, this.props.rows);
+        }
+
+        this.setTerminalTheme(this.props.theme || 'black-theme');
+        this.terminal.focus();
+    }
 
+    unmountTerminal(state: TerminalState) {
+        this.terminalRef.current?.removeChild(state.element);
+        // This makes sure that the terminal will not cause its new
+        // container to grow when it is reattached later.
+        if (!this.props.rows)
+            this.resizeTerminal(80, 1);
+    }
+
+    componentDidMount() {
+        this.mountTerminal(this.terminal_state);
         if (!this.props.rows) {
             window.addEventListener('resize', this.onWindowResize);
             this.onWindowResize();
         }
-        this.setTerminalTheme(this.props.theme || 'black-theme');
-        this.terminal.focus();
     }
 
-    resizeTerminal(cols, rows) {
+    resizeTerminal(cols: number, rows: number) {
         this.terminal.resize(cols, rows);
-        this.props.channel.control({
-            window: {
-                rows,
-                cols
-            }
-        });
+        if (this.terminal_state.channel) {
+            this.terminal_state.channel.control({
+                window: {
+                    rows,
+                    cols
+                }
+            } as cockpit.JsonObject as cockpit.ControlMessage);
+        }
     }
 
-    componentDidUpdate(prevProps, prevState) {
-        if (prevProps.fontSize !== this.props.fontSize) {
+    componentDidUpdate(prevProps: TerminalComponentProps, prevState: TerminalComponentState) {
+        if (this.props.state && prevProps.state !== this.props.state) {
+            cockpit.assert(!this.props.channel);
+            cockpit.assert(prevProps.state);
+            this.unmountTerminal(prevProps.state);
+            this.terminal_state = this.props.state;
+            this.mountTerminal(this.terminal_state);
+            if (!this.props.cols || !this.props.rows)
```

Comment:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-21680-240581e9-20250828-085911-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)